File: src-tauri/src/commands/story_system.rs
//! Story System commands

use tauri::{State, AppHandle};
use crate::db::DbPool;
use crate::error::AppError;

// ==================== Story System Commands ====================

#[tauri::command(rename_all = "snake_case")]
pub fn create_master_setting(
    story_id: String,
    genre: String,
    core_tone: String,
    pacing_strategy: String,
    anti_patterns: Vec<String>,
    world_rules: Vec<String>,
    pool: State<'_, DbPool>,
    app: AppHandle,
) -> Result<crate::db::StoryContract, AppError> {
    let pool = pool.inner().clone();
    let engine = crate::story_system::StorySystemEngine::new(pool);
    let result = engine.create_master_setting(
        &story_id, &genre, &core_tone, &pacing_strategy, &anti_patterns, &world_rules
    )
    .map_err(AppError::from)?;
    let _ = crate::state_sync::StateSync::emit_data_refresh(&app, Some(&story_id), "storyContracts");
    Ok(result)
}


#[tauri::command(rename_all = "snake_case")]
pub fn create_chapter_contract(
    story_id: String,
    chapter_number: i32,
    goal: String,
    must_cover_nodes: Vec<String>,
    forbidden_zones: Vec<String>,
    time_anchor: Option<String>,
    chapter_span: Option<String>,
    pool: State<'_, DbPool>,
    app: AppHandle,
) -> Result<crate::db::StoryContract, AppError> {
    let pool = pool.inner().clone();
    let engine = crate::story_system::StorySystemEngine::new(pool);
    let result = engine.create_chapter_contract(
        &story_id, chapter_number, &goal, &must_cover_nodes, &forbidden_zones,
        time_anchor.as_deref(), chapter_span.as_deref()
    )
    .map_err(AppError::from)?;
    let _ = crate::state_sync::StateSync::emit_data_refresh(&app, Some(&story_id), "storyContracts");
    Ok(result)
}


#[tauri::command(rename_all = "snake_case")]
pub fn get_contract_tree(story_id: String, pool: State<'_, DbPool>) -> Result<crate::story_system::ContractTree, AppError> {
    let pool = pool.inner().clone();
    let engine = crate::story_system::StorySystemEngine::new(pool);
    engine.get_contract_tree(&story_id)
        .map_err(AppError::from)
}


#[tauri::command(rename_all = "snake_case")]
pub fn get_runtime_contract(
    story_id: String,
    chapter_number: i32,
    pool: State<'_, DbPool>,
) -> Result<crate::story_system::RuntimeContract, AppError> {
    let pool = pool.inner().clone();
    let engine = crate::story_system::StorySystemEngine::new(pool);
    engine.get_runtime_contract(&story_id, chapter_number)
        .map_err(AppError::from)
}


#[tauri::command(rename_all = "snake_case")]
pub fn init_chapter_commit(
    story_id: String,
    scene_id: Option<String>,
    chapter_id: Option<String>,
    chapter_number: i32,
    pool: State<'_, DbPool>,
    app: AppHandle,
) -> Result<crate::db::SceneCommit, AppError> {
    let pool = pool.inner().clone();
    let service = crate::story_system::SceneCommitService::new(pool);
    let result = service.init_commit(&story_id, scene_id.as_deref(), chapter_id.as_deref(), chapter_number)
        .map_err(AppError::from)?;
    let _ = crate::state_sync::StateSync::emit_data_refresh(&app, Some(&story_id), "sceneCommits");
    Ok(result)
}


#[tauri::command(rename_all = "snake_case")]
pub fn get_chapter_commits(story_id: String, pool: State<'_, DbPool>) -> Result<Vec<crate::db::SceneCommit>, AppError> {
    let pool = pool.inner().clone();
    let repo = crate::db::SceneCommitRepository::new(pool);
    repo.get_by_story(&story_id).map_err(AppError::from)
}


#[tauri::command(rename_all = "snake_case")]
pub fn list_genesis_runs(limit: Option<i64>, pool: State<'_, DbPool>) -> Result<Vec<crate::db::GenesisRun>, AppError> {
    let pool = pool.inner().clone();
    let repo = crate::db::GenesisRunRepository::new(pool);
    repo.list_all(limit.unwrap_or(100)).map_err(AppError::from)
}


#[tauri::command(rename_all = "snake_case")]
pub fn get_genesis_run(id: String, pool: State<'_, DbPool>) -> Result<Option<crate::db::GenesisRun>, AppError> {
    let pool = pool.inner().clone();
    let repo = crate::db::GenesisRunRepository::new(pool);
    repo.get_by_id(&id).map_err(AppError::from)
}


#[tauri::command(rename_all = "snake_case")]
pub fn get_latest_style_snapshot(story_id: String, pool: State<'_, DbPool>) -> Result<Option<crate::db::models::StyleSnapshot>, AppError> {
    let pool = pool.inner().clone();
    let repo = crate::db::StyleSnapshotRepository::new(pool);
    repo.get_latest_by_story(&story_id).map_err(AppError::from)
}

File: src-frontend/src/services/api/storySystem.ts
import { loggedInvoke } from './core';
// ==================== v6.0.0: Story System ====================

export interface StoryContract {
  id: string;
  story_id: string;
  contract_type: string;
  contract_json: string;
  version: number;
  created_at: string;
  updated_at: string;
}

export interface ChapterCommit {
  id: string;
  story_id: string;
  scene_id: string | null;
  chapter_id: string | null;
  chapter_number: number;
  status: string;
  outline_snapshot_json: string | null;
  review_result_json: string | null;
  fulfillment_result_json: string | null;
  accepted_events_json: string | null;
  state_deltas_json: string | null;
  entity_deltas_json: string | null;
  summary_text: string | null;
  dominant_strand: string | null;
  projection_status_json: string | null;
  created_at: string;
}

export interface ContractTree {
  master_setting: StoryContract | null;
  volumes: Record<string, StoryContract>;
  chapters: Record<string, StoryContract>;
  reviews: Record<string, StoryContract>;
}

export interface RuntimeContract {
  master_setting: StoryContract;
  chapter_contract: StoryContract | null;
}

export const createMasterSetting = (params: {
  story_id: string;
  genre: string;
  core_tone: string;
  pacing_strategy: string;
  anti_patterns: string[];
  world_rules: string[];
}) => loggedInvoke<StoryContract>('create_master_setting', params);

export const createChapterContract = (params: {
  story_id: string;
  chapter_number: number;
  goal: string;
  must_cover_nodes: string[];
  forbidden_zones: string[];
  time_anchor?: string;
  chapter_span?: string;
}) => loggedInvoke<StoryContract>('create_chapter_contract', params);

export const getContractTree = (storyId: string) =>
  loggedInvoke<ContractTree>('get_contract_tree', { story_id: storyId });

export const getRuntimeContract = (storyId: string, chapterNumber: number) =>
  loggedInvoke<RuntimeContract>('get_runtime_contract', {
    story_id: storyId,
    chapter_number: chapterNumber,
  });

export const initChapterCommit = (
  storyId: string,
  chapterNumber: number,
  sceneId?: string,
  chapterId?: string
) =>
  loggedInvoke<ChapterCommit>('init_chapter_commit', {
    story_id: storyId,
    chapter_number: chapterNumber,
    scene_id: sceneId,
    chapter_id: chapterId,
  });

// W2-B5: apply_chapter_commit 已改为 update_chapter 成功后自动触发（30s debounce）
// 前端不再显式调用，保留 get_chapter_commits 用于展示 commit 历史
export const getChapterCommits = (storyId: string) =>
  loggedInvoke<ChapterCommit[]>('get_chapter_commits', { story_id: storyId });
